Watch if operators are in breach of the threshold

samigabor opened the pull request:
## Summary

[Tries to find Operator contracts that are in breach with the update rules, to cash in rewards. Periodically select an Operator contract and inspect if total diff in accumulated rewards is over the threshold. If so, auto-withdraw them from sponsorship to operator contract and get a share of the operator's cut, for doing his job.](https://linear.app/streamr/issue/ETH-537/operatorvaluebreachwatcher)

## Changes

Watcher implementation and maintain operator update to auto-withdraw.
Other fastchain related fixes.

samigabor committed:
feat: watch if operators are in breach of the threshold

samigabor committed:
Merge branch 'streamr-1.0' into NET-537_operator_value_breach_watcher

samigabor committed:
feat: update watcher service and helper

samigabor committed:
feat: update watcher service, update maintain value service, refactor helper

samigabor committed:
Merge branch 'streamr-1.0' into NET-537_operator_value_breach_watcher

samigabor committed:
fix: token interface import

samigabor committed:
fix: loop over the sorted sponsorships not all to sum the earnings

jtakalai committed:
doc: commenting and names cleanup

jtakalai committed:
feat: change MaintainOperatorValueService constructor argument to "safety limit" instead of penalty limit

This way the operator can choose their "safety level" and not care about the contract's own penalty limit that can't be changed anyway

samigabor committed:
feat: make sponsorship specific config params optional

samigabor committed:
feat: add sponsorship earnings issuance param

samigabor committed:
test: happy path for withdraw earnings from sponsorships

jtakalai committed:
test: rewrote MaintainOperatorValueService test after the unwithdrawn earnings simplification

samigabor committed:
Merge branch 'streamr-1.0' into NET-537_operator_value_breach_watcher

samigabor committed:
fix: use signer for operator contract setup in helper, test against own operator address (will be removed and used a new dev env), add wait for condition after start

samigabor committed:
feat: filter out own operator before returning a random operator

samigabor committed:
refactor: add helper method to sum earnings

samigabor committed:
test: fix maintain operator value test case, refactor watcher test case

jtakalai committed:
build: update network-contracts dependency

jtakalai committed:
refactor: move logic from Helper to Service

keep smart contract interactions in Helper

jtakalai committed:
WIP: fixing the operator value watcher test

why is streamExists=false right after it's created?!

jtakalai committed:
test: version of OperatorValueBreachWatcher.test.ts that works with dev-chain-fast

jtakalai committed:
Merge remote-tracking branch 'origin/streamr-1.0' into NET-537_operator_value_breach_watcher

# Conflicts:
#	packages/broker/src/plugins/operator/OperatorPlugin.ts

jtakalai committed:
feat: operator value related tests passing using dev-chain-fast

of course dev-chain-fast related changes are left out of this commit. This is to see if it passes in the CI that uses the old dev-docker parity chain

jtakalai committed:
fix: test provider

use the smartContractUtils for more things. Had to workaround for dev-chain-fast (that I'm using locally now)

jtakalai committed:
test: fix sidechain connection

jtakalai committed:
Merge branch 'streamr-1.0' into NET-537_operator_value_breach_watcher

jtakalai committed:
build: clear package changes from this PR branch

jtakalai committed:
doc: add debugging

jtakalai committed:
build: update network-contracts

envdeployer has changed at least, so compilation requires 4.3.0

jtakalai committed:
feat: add checkValueIntervalMs to MaintainOperatorValueService constructor

jtakalai committed:
feat: breach watcher should trigger withdraw in the target Operator contract

...not in their own Operator contract

jtakalai committed:
build: switch to use the fast chain for operator plugin tests

jtakalai committed:
test: adjust timeouts

samigabor committed:
build: update streamr config to latest

samigabor committed:
test: watcher in  two fazes since it can not query contracts on newly deployed environments

First test the get random operator method
Then override get random function to return a known operator

samigabor committed:
refactor: use config contracts instead of deploying a new dev env and new contracts

samigabor committed:
refactor: remove commented code, reorder items

samigabor committed:
test: check the returned random operator is deployed by the operator factory

samigabor committed:
Merge branch 'streamr-1.0' into NET-537_operator_value_breach_watcher

samigabor committed:
Merge branch 'streamr-1.0' into NET-537_operator_value_breach_watcher

samigabor committed:
fix: remove unused imports

samigabor committed:
build: update config version

samigabor committed:
fix: change graph port to index the fastchain

samigabor committed:
Merge branch 'streamr-1.0' into NET-537_operator_value_breach_watcher

samigabor committed:
fix: eslint quotes error

samigabor committed:
Merge branch 'streamr-1.0' into NET-537_operator_value_breach_watcher

samigabor committed:
fix: eslint quotes error

samigabor committed:
fix: update hard-coded operator address for the new chain image

harbu commented on .github/workflows/validate.yml:
```diff
@@ -110,6 +110,14 @@ jobs:
+      package: broker
+      docker-services: cassandra init-keyspace parity-sidechain-node0 graph-deploy-streamregistry-subgraph deploy-network-subgraphs deploy-network-subgraphs-fastchain
+      run-entry-point: true
+      command: npm run test-operator-plugins
```
fast-chain is now in use in normal broker-e2e tests, so this can be removed?

samigabor replied:
Yes, removed it.

harbu commented on packages/broker/package.json:
```diff
@@ -22,14 +22,15 @@
     "test-sequential": "jest --forceExit --maxWorkers=1 test/sequential # always run sequential tests with maxWorkers=1",
-    "test-integration": "jest --forceExit test/integration && npm run test-sequential"
+    "test-integration": "jest --testPathIgnorePatterns='test/integration/plugins/operator/' --forceExit test/integration && npm run test-sequential",
+    "test-operator-plugins": "jest --forceExit test/integration/plugins/operator"
```
fast-chain is now in use in normal broker-e2e tests, so this can be reverted?

samigabor replied:
yes, reverted

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,127 @@
+        const operators = await this.getOperatorAddresses(latestBlock, queryFilter)
+        // filter out my own operator
+        const operatorAddresses = operators.filter((id) => id !== this.config.operatorContractAddress)
+        logger.info(`Found ${operatorAddresses.length} operators`, { operatorAddresses })
```
Let's keep details at `debug` level as our end-users don't need to know about this details unless they want to diagnose an issue and enable `LOG_LEVEL=debug`.

```suggestion
        logger.debug(`Found ${operatorAddresses.length} operators`, { operatorAddresses })
```

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,127 @@
+            .filter((sponsorship) => sponsorship.earnings.gte(minSponsorshipEarningsWei))
+            .sort((a, b) => Number(b.earnings.sub(a.earnings).toBigInt())) // TODO: after Node 20, use .toSorted() instead
+            .slice(0, this.config.maxSponsorshipsCount) // take all if maxSponsorshipsCount is undefined
+        const sponsorshipAddresses = sponsorships.map((sponsorship) => sponsorship.address as EthereumAddress)
```
If you use the `EthereumAddress` type it is assumed that it is (a) a valid address and (b) in lower-case. Here the first one most likely always holds since it is read from a smart contract but the second one not.

So we should avoid casting here and instead use the helper function `toEthereumAddress`:

```suggestion
        const sponsorshipAddresses = sponsorships.map((sponsorship) => toEthereumAddress(sponsorship.address))
```

samigabor replied:
Changed

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,127 @@
+            return {
+                query: `
+                    {
+                        operators${queryFilter} {
```
Parameter `queryFilter` left blank in actual uses of method. I would remove parameter.
```suggestion
    private async getOperatorAddresses(requiredBlockNumber: number): Promise<EthereumAddress[]> {
        const createQuery = () => {
            return {
                query: `
                    {
                        operators {
```

jtakalai replied:
there should be pagination coming up (for the random selection), so the queryFilter is probably needed

teogeb replied:
The pagination is automatically handled if we inject the the parameters of `createQuery` (`lastId: string, pageSize: number`) to line `101`. (see e.g. https://github.com/streamr-dev/network/blob/f782bb9155e2d399fe5424d56e9a6b411945bc27/packages/broker/src/plugins/operator/MaintainTopologyHelper.ts#L107)

samigabor replied:
removed param

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,127 @@
+        }
+        const parseItems = (response: any) => {
+            if (!response.operators) {
+                logger.error('Unable to find operators in The Graph')
```
In a sense a "valid" situation (given no operators around) and not an error. I'd remove logging.
```suggestion
```

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,127 @@
+            return response.operators
+        }
+        this.theGraphClient.updateRequiredBlockNumber(requiredBlockNumber)
+        const queryResult = this.theGraphClient.queryEntities<any>(createQuery, parseItems)
```
```suggestion
        const queryResult = this.theGraphClient.queryEntities<{ id: string }>(createQuery, parseItems)
```

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
+    private penaltyLimitFraction?: bigint
     private readonly helper: MaintainOperatorValueHelper
     private readonly abortController: AbortController
+    private readonly checkIntervalMs: number
```
(nitpick) To be in line with rest of code base, we've usually used "inMs" suffix. 
```suggestion
    private readonly checkIntervalInMs: number
```

samigabor replied:
Changed

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
+        this.checkIntervalMs = checkValueIntervalMs
+    }
+
+    async checkMyUnwithdrawnEarnings(): Promise<void> {
```
A private helper method.
```suggestion
    private async checkMyUnwithdrawnEarnings(): Promise<void> {
```

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
     }
 
     async start(): Promise<void> {
+        this.penaltyLimitFraction = await this.helper.getPenaltyLimitFraction()
```
(nitpick) Let's move `start` method right after constructor for clarity. (Since everything else is a private helper method)

samigabor replied:
Moved

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
-            () => this.checkValue().catch((err) => {
-                logger.warn('Encountered error while checking value', { err })
+            () => this.checkMyUnwithdrawnEarnings().catch((err) => {
+                logger.warn('Encountered error during checkUnwithdrawnEarnings', { err })
```
We could even elevate this to `error` level since this isn't a typical error scenario, right?
```suggestion
                logger.error('Encountered error while checking unwithdrawn earnings', { err })
```

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
-    }
-
     async stop(): Promise<void> {
         logger.info('Stop')
```
Not really needed imo
```suggestion
```

samigabor replied:
removed

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
+    }
+
+    async checkMyUnwithdrawnEarnings(): Promise<void> {
+        logger.info('Check unwithdrawn earnings % of total value, and check if it\'s above the safe threshold')
```
This is good to log imo (since it happens only hourly), can we make the log message a bit more succinct so it is clear to the end-user what is happening?

Maybe something like?
```suggestion
        logger.info('Check whether it is time to withdraw my earnings')
```

samigabor replied:
Changed

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
+    async checkMyUnwithdrawnEarnings(): Promise<void> {
+        logger.info('Check unwithdrawn earnings % of total value, and check if it\'s above the safe threshold')
+        const { fraction, sponsorshipAddresses } = await this.helper.getMyUnwithdrawnEarnings()
+        const safeUnwithdrawnEarningsFraction = this.penaltyLimitFraction! * this.withdrawLimitSafetyFraction / BigInt(1e18)
```
I noticed we multiplied `withdrawLimitSafetyFraction` with `1e18` in constructor and here we divide by `1e18` reversing that multiplication? So would it be equivalent just to remove the multiplication (in the constructor) and division (here)?
 
```suggestion
        const safeUnwithdrawnEarningsFraction = this.penaltyLimitFraction! * this.withdrawLimitSafetyFraction / BigInt(1e18)
```

samigabor replied:
Multiplication needed in the constructor in order to convert it to BigInt. 
And this way, it also reflects the wei value you have in the contact

samigabor replied:
Left the multiplication, but extracted the one ether value into a const (as suggested below)

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
+        logger.info('Check unwithdrawn earnings % of total value, and check if it\'s above the safe threshold')
+        const { fraction, sponsorshipAddresses } = await this.helper.getMyUnwithdrawnEarnings()
+        const safeUnwithdrawnEarningsFraction = this.penaltyLimitFraction! * this.withdrawLimitSafetyFraction / BigInt(1e18)
+        logger.info(` -> is ${fraction / BigInt(1e16)}% > ${safeUnwithdrawnEarningsFraction / BigInt(1e16)}% ?`)
```
Development-time debugging log? If so remove, or if you feel it is useful to have around maybe make into `trace` level, e.g.

```suggestion
        logger.trace(` -> is ${fraction / BigInt(1e16)}% > ${safeUnwithdrawnEarningsFraction / BigInt(1e16)}% ?`)
```

samigabor replied:
Changed to trace

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
+        const safeUnwithdrawnEarningsFraction = this.penaltyLimitFraction! * this.withdrawLimitSafetyFraction / BigInt(1e18)
+        logger.info(` -> is ${fraction / BigInt(1e16)}% > ${safeUnwithdrawnEarningsFraction / BigInt(1e16)}% ?`)
+        if (fraction > safeUnwithdrawnEarningsFraction) {
+            logger.info('Withdrawing earnings from sponsorships', { sponsorshipAddresses })
```
```suggestion
            logger.info('Withdraw earnings from sponsorships', { sponsorshipAddresses })
```

harbu commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
+        if (fraction > safeUnwithdrawnEarningsFraction) {
+            logger.info('Withdrawing earnings from sponsorships', { sponsorshipAddresses })
+            await this.helper.withdrawMyEarningsFromSponsorships(sponsorshipAddresses)
+        }
```
May be useful to let the user know also when we decide not to withdraw.

```suggestion
            await this.helper.withdrawMyEarningsFromSponsorships(sponsorshipAddresses)
        } else {
            logger.info('Skip withdrawing earnings', { fraction, safeUnwithdrawnEarningsFraction })
        }
```

samigabor replied:
Added

teogeb commented on .github/workflows/validate.yml:
```diff
@@ -110,6 +110,14 @@ jobs:
       docker-services: cassandra init-keyspace dev-chain-fast deploy-network-subgraphs-fastchain
       run-entry-point: true
       command: npm run test-integration
+  broker-e2e-fast:
```
The streamr-1.0 branch now uses fastchain, so we don't need this if you merge latest from that branch

samigabor replied:
Removed

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+    private penaltyLimitFractionCached?: bigint
+    private readonly abortController: AbortController
+    
+    readonly helper: MaintainOperatorValueHelper
```
```suggestion
    // public access modifier for tests 
    readonly helper: MaintainOperatorValueHelper
```

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    async stop(): Promise<void> {
+        logger.info('Stop')
```
```suggestion
```

samigabor replied:
Removed

teogeb commented on packages/broker/package.json:
```diff
@@ -22,14 +22,15 @@
     "test-unit": "jest test/unit",
     "test-sequential": "jest --forceExit --maxWorkers=1 test/sequential # always run sequential tests with maxWorkers=1",
-    "test-integration": "jest --forceExit test/integration && npm run test-sequential"
+    "test-integration": "jest --testPathIgnorePatterns='test/integration/plugins/operator/' --forceExit test/integration && npm run test-sequential",
```
It would be good to run all integration tests with `test-integration`.

Is there a need for this change as we can run single test or subset of tests from command line with `npx`:
`npx jest test/integration/plugins/operator/OperatorPlugin.test.ts`

jtakalai replied:
the reason for splitting operator tests out was to run them with fast-chain. So if now all `test-integration` already work with fast-chain, then this split isn't needed.

samigabor replied:
Removed the split.

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+        return this.penaltyLimitFractionCached
+    }
+
+    async start(): Promise<void> {
```
(nitpick): Move this method right after constructor.

samigabor replied:
Moved

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+    async start(): Promise<void> {
+        await scheduleAtInterval(
+            () => this.checkRandomUnwithdrawnEarnings().catch((err) => {
+                logger.warn('Encountered error while watching operators', { err })
```
```suggestion
                logger.warn('Encountered error', { err })
```

samigabor replied:
Updated

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+        this.abortController = new AbortController()
+    }
+
+    async checkRandomUnwithdrawnEarnings(): Promise<void> {
```
```suggestion
    private async checkRandomUnwithdrawnEarnings(): Promise<void> {
```

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+
+    async checkRandomUnwithdrawnEarnings(): Promise<void> {
+        const randomOperatorAddress = await this.helper.getRandomOperator()
+        logger.info('Checking unwithdrawn earnings', { randomOperatorAddress })
```
```suggestion
        logger.info('Check unwithdrawn earnings', { randomOperatorAddress })
```

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+        return this.checkUnwithdrawnEarningsOf(randomOperatorAddress)
+    }
+
+    async checkUnwithdrawnEarningsOf(targetOperatorAddress: EthereumAddress): Promise<void> {
```
```suggestion
    private async checkUnwithdrawnEarningsOf(targetOperatorAddress: EthereumAddress): Promise<void> {
```

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    async checkUnwithdrawnEarningsOf(targetOperatorAddress: EthereumAddress): Promise<void> {
+        logger.info('Check unwithdrawn earnings and check if they are above the safe threshold')
```
We already log an info statement the the caller `checkRandomUnwithdrawnEarnings`. A bit redundant here (from end-user perspective).
```suggestion
```

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+        logger.info('Check unwithdrawn earnings and check if they are above the safe threshold')
+        const { fraction, sponsorshipAddresses } = await this.helper.getUnwithdrawnEarningsOf(targetOperatorAddress)
+        const limit = await this.getPenaltyLimitFraction()
+        logger.info(` -> is ${fraction} > ${limit}?`)
```
Looks like a development-time debug statement. Remove, or move to trace level:
```suggestion
        logger.trace(` -> is ${fraction} > ${limit}?`)
```

samigabor replied:
Changed

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+        const limit = await this.getPenaltyLimitFraction()
+        logger.info(` -> is ${fraction} > ${limit}?`)
+        if (fraction > limit) {
+            logger.info('Withdrawing earnings from sponsorships', { sponsorshipAddresses })
```
```suggestion
            logger.info('Withdraw earnings from sponsorships (target operator value in breach)', { targetOperatorAddress, sponsorshipAddresses, fraction, limit })
```

samigabor replied:
Updated

harbu commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    async getPenaltyLimitFraction(): Promise<bigint> {
```
```suggestion
    private async getPenaltyLimitFraction(): Promise<bigint> {
```

harbu commented on packages/broker/test/integration/plugins/operator/OperatorValueBreachWatcher.test.ts:
```diff
@@ -0,0 +1,128 @@
+            minSponsorshipEarnings: 1 // full tokens
+        }
+        return { operatorWallet, operatorContract, operatorConfig }
+    }
```
Can we use pre-existing `setupOperatorContract` to avoid repeating code?

samigabor replied:
Couldn't use that one because `{ operatorSharePercent: 10 }` param was needed

harbu replied:
I would solve this by adding the possibility to pass that param into setupOperatorContract; see https://github.com/streamr-dev/network/pull/1763 for my solution idea.

samigabor replied:
Reviewed it. Thanks

harbu commented on packages/broker/test/integration/plugins/operator/OperatorValueBreachWatcher.test.ts:
```diff
@@ -0,0 +1,128 @@
+        // TODO: use createClient once configs allow it. For now I'm creating the stream directly using the contract
+        // const client = createClient(STREAM_CREATION_KEY)
+        // streamId = (await createTestStream(client, module)).id
+        // await client.destroy()
```
You can use the `client` to create the stream now since we are using fast-chain by default in client now.

samigabor replied:
Changed it

harbu commented on packages/broker/test/integration/plugins/operator/OperatorValueBreachWatcher.test.ts:
```diff
@@ -0,0 +1,128 @@
+        token = new Contract(fastChainConfig.contracts.DATA, tokenABI) as unknown as TestToken
+    }, 60 * 1000)
+
+    it('can find a random operator, excluding himself', async () => {
```
This is really testing the method `getRandomOperator` of class `MaintainOperatorValueHelper`.

I would create a new integration test `MaintainOperatorValueHelper.test.ts" and move this test there.

harbu replied:
I noticed this wasn't moved. I would move it as done here https://github.com/streamr-dev/network/pull/1763

samigabor replied:
Sorry about this one. I see now I missed it..

harbu commented on packages/broker/test/integration/plugins/operator/MaintainOperatorValueService.test.ts:
```diff
@@ -1,125 +1,108 @@
 
-// test is outdated, is completely rewritten and will be merged with PR #1629 
-describe.skip('MaintainOperatorValueService', () => {
+async function getTotalUnwithdrawnEarnings(operatorContract: Operator): Promise<bigint> {
```
Same function duplicated in OperatorValueBreachWatcher.test.ts? Extract into common helper function and use in both to avoid code repetition

samigabor replied:
Extracted

harbu commented on packages/broker/test/integration/plugins/operator/MaintainOperatorValueService.test.ts:
```diff
@@ -1,125 +1,108 @@
         }
-        return totalDiff
+        return { operatorWallet, operatorContract, operatorConfig }
     }
```
Can we use pre-existing `setupOperatorContract` for this? (To avoid code repetition)

samigabor replied:
Couldn't use that one because { operatorSharePercent: 10 } param was needed

harbu commented on packages/broker/test/integration/plugins/operator/MaintainOperatorValueService.test.ts:
```diff
@@ -1,125 +1,108 @@
+        // TODO: use createClient once configs allow it. For now I'm creating the stream directly using the contract
+        // const client = createClient(STREAM_CREATION_KEY)
+        // streamId = (await createTestStream(client, module)).id
+        // await client.destroy()
```
fast-chain is enabled in client now in streamr-1.0 branch. So you can use `client` to create streams.

samigabor replied:
Updated.

jtakalai committed:
build: remove split integration tests (all are fast-chain after merge from streamr-1.0)

jtakalai committed:
Merge remote-tracking branch 'origin/streamr-1.0' into NET-537_operator_value_breach_watcher

jtakalai committed:
fix: use converter instead of coercion

jtakalai committed:
fix: delete _meta from query

samigabor committed:
fix: update hard-coded preloaded operator contract address according to new fast-chain image build

samigabor committed:
fix: guard against division by zero on penalty fraction calculation and earnings unwithdraw when there are no operators deployed

samigabor committed:
refactor: use client to generate test stream

teogeb commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,124 @@
+        const queryFilter = '' // e.g. (first: 10, orderBy: poolValue, orderDirection: desc)
+        const operators = await this.getOperatorAddresses(latestBlock, queryFilter)
+        // filter out my own operator
+        const operatorAddresses = operators.filter((id) => toEthereumAddress(id) !== toEthereumAddress(this.config.operatorContractAddress))
```
both of the comparison operands have already the type of `EthereumAddress`

```suggestion
        const operatorAddresses = operators.filter((id) => id !== this.config.operatorContractAddress)
```

teogeb commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,124 @@
+        const operatorAddresses = operators.filter((id) => toEthereumAddress(id) !== toEthereumAddress(this.config.operatorContractAddress))
+        logger.info(`Found ${operatorAddresses.length} operators`, { operatorAddresses })
+        const randomIndex = Math.floor(Math.random() * operatorAddresses.length)
+        return operatorAddresses[randomIndex]
```
`import sample from 'lodash/sample'` and

```suggestion
        return sample(operatorAddresses)
```

samigabor replied:
Updated

teogeb commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
-    constructor(config: OperatorServiceConfig, penaltyLimitFraction = BigInt(0)) {
-        this.penaltyLimitFraction = penaltyLimitFraction
+    constructor(config: OperatorServiceConfig, withdrawLimitSafetyFraction = 0.5, checkValueIntervalMs = DEFAULT_CHECK_VALUE_INTERVAL_MS) {
+        this.withdrawLimitSafetyFraction = BigInt(withdrawLimitSafetyFraction * 1e18)
```
Use `ONE_ETHER` from the helper (also at line 26)

samigabor replied:
Updated

teogeb commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,81 +1,49 @@
 
-    constructor(config: OperatorServiceConfig, penaltyLimitFraction = BigInt(0)) {
-        this.penaltyLimitFraction = penaltyLimitFraction
+    constructor(config: OperatorServiceConfig, withdrawLimitSafetyFraction = 0.5, checkValueIntervalMs = DEFAULT_CHECK_VALUE_INTERVAL_MS) {
```
The default value of `withdrawLimitSafetyFraction` could be a constant, similarly to `checkValueIntervalMs`.

samigabor replied:
Updated

teogeb commented on packages/broker/src/plugins/operator/OperatorPlugin.ts:
```diff
@@ -50,7 +54,9 @@ export class OperatorPlugin extends Plugin<OperatorPluginConfig> {
-            signer: Wallet.createRandom().connect(provider)
+            signer: Wallet.createRandom().connect(provider),
+            maxSponsorshipsCount: 20, // max number of sponsorships to loop over before tx reverts
+            minSponsorshipEarnings: 1 // token value, not wei
```
Introduce constants for these?

samigabor replied:
Updated

teogeb commented on packages/broker/src/plugins/operator/OperatorPlugin.ts:
```diff
@@ -27,6 +28,8 @@ export interface OperatorServiceConfig {
     signer: Signer
     operatorContractAddress: EthereumAddress
     theGraphUrl: string
+    maxSponsorshipsCount?: number
```
Maybe a more specific name if this limits just the operator value maintenance, not anything else?

jtakalai replied:
maybe maxSponsorshipsInWithdraw

samigabor replied:
Updated

teogeb commented on packages/broker/test/integration/plugins/operator/MaintainOperatorValueService.test.ts:
```diff
@@ -1,125 +1,96 @@
     }
 
     beforeAll(async () => {
+        provider = new JsonRpcProvider(`http://${STREAMR_DOCKER_DEV_HOST}:8547`)
```
Url from `chainConfig.rpcEndpoints[0].url`

samigabor replied:
Updated

teogeb commented on packages/broker/test/integration/plugins/operator/MaintainOperatorValueService.test.ts:
```diff
@@ -1,125 +1,96 @@
+
+        const poolValueBeforeWithdraw = await operatorContract.getApproximatePoolValue()
+
+        // safe threshold is 5 DATA, that's when withdraw happens
```
Which calculation or config defines that value of `5`? (Same for value `3` at lines `87`-`88`)

jtakalai replied:
TODO: add more comments.

5 DATA = 50% of 10 DATA, 50% is the "safe threshold" (we don't want to wait all the way to the limit, lest we be late)
10 DATA = 10% of 100 DATA, 10% is the penalty limit (sum of unwithdrawn earnings may not exceed it) 
100 DATA = pool value

samigabor replied:
Added more comments

teogeb commented on packages/broker/test/integration/plugins/operator/OperatorValueBreachWatcher.test.ts:
```diff
@@ -0,0 +1,121 @@
+        const randomOperatorAddress = await operatorValueBreachWatcher.helper.getRandomOperator()
+        if (randomOperatorAddress === undefined) {
+            throw new Error('No random operator found')
+        }
```
```suggestion
        expect(randomOperatorAddress).toBeDefined()
```

jtakalai replied:
one of the points of having that comparison is making it clear to TypeScript that `randomOperatorAddress` can not be undefined. TS can't infer it from `expect`.

samigabor replied:
Left it as-is for now

teogeb commented on packages/broker/test/integration/plugins/operator/OperatorValueBreachWatcher.test.ts:
```diff
@@ -0,0 +1,121 @@
+        // check it's a valid operator, deployed by the OperatorFactory
+        const adminWallet = new Wallet(STREAM_CREATION_KEY, provider)
+        const operatorFactory = new Contract(fastChainConfig.contracts.OperatorFactory, operatorFactoryABI, adminWallet) as unknown as OperatorFactory
+        const isDeployedByFactory = await operatorFactory.deploymentTimestamp(randomOperatorAddress)
```
Is the variable name correct?

jtakalai replied:
yes, if it has a deployment timestamp then it was deployed by that factory. I guess it could be cast to bool by adding `.gt(0)` to the end

samigabor replied:
Updated

samigabor committed:
Update packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
Update packages/broker/src/plugins/operator/MaintainOperatorValueService.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
refactor: remove unused query filter

samigabor committed:
Update packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
Update packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
refactor: fix var naming

samigabor committed:
Update packages/broker/src/plugins/operator/MaintainOperatorValueService.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
refactor: move start method after constructor

samigabor committed:
refactor: remove stop log

samigabor committed:
refactor: update log message

samigabor committed:
refactor: change log level for dev

samigabor committed:
refactor: add log to let user know when withdrawing is skipped

samigabor committed:
refactor: update log message

samigabor committed:
Update packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
refactor: move start method after constructor

samigabor committed:
refactor: update log message

samigabor committed:
Update packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
Update packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
Update packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
Update packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
refactor: update log message

samigabor committed:
Update packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts

Co-authored-by: Eric Andrews <eric.andrews@streamr.com>

samigabor committed:
Update packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts

Co-authored-by: Teo Gebhard <teo@streamr.com>

samigabor committed:
refactor: use lodash to get random operator

samigabor committed:
refactor: extract common method into shared file

samigabor committed:
refactor: extract one ether value into const

samigabor committed:
refactor: extract default withdraw limit safety fraction value into const

samigabor committed:
refactor: extract default values for sponsorships filtering into constants

samigabor committed:
refactor: use more descriptive names for sponsorships filtering variables

samigabor committed:
docs: add comments to better explain the threshold

samigabor committed:
refactor: use rpc url from config, rename fastChainConfig to fastChain for consistency

samigabor committed:
refactor: change type and assertion on random operator deployment status

samigabor committed:
Merge branch 'streamr-1.0' into NET-537_operator_value_breach_watcher

harbu reviewed (commented):
Most things looking very good now 👍 , I had couple of suggestions more, I thought easiest maybe just to write them down as code see here: https://github.com/streamr-dev/network/pull/1763 and let me know what you think.

Commit:
Pr suggestion 1 (#1763)

## Summary

Two suggestions to PR https://github.com/streamr-dev/network/pull/1629

1) Use `setupOperatorContract` in introduced integration tests to avoid
repetition.

2) Move `can find a random operator, excluding himself` under file
MaintainOperatorValueHelper.test.ts since it is really testing its
behavior.

teogeb commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,122 @@
+                    `
+            }
+        }
+        const parseItems = (response: any) => {
```
Stricter type?

samigabor replied:
Updated

teogeb commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,122 @@
+        this.theGraphClient.updateRequiredBlockNumber(requiredBlockNumber)
+        const queryResult = this.theGraphClient.queryEntities<{ id: string }>(createQuery, parseItems)
+
+        const operatorAddresses: EthereumAddress[] = []
+        for await (const operator of queryResult) {
+            operatorAddresses.push(toEthereumAddress(operator.id))
+        }
+        return operatorAddresses
```
```suggestion
        return queryResult.map((operator) => toEthereumAddress(operator.id))
```

samigabor replied:
Hmm, that would not work though.. queryResult is an AsyncGenerator..

teogeb commented on packages/broker/src/plugins/operator/MaintainOperatorValueService.ts:
```diff
@@ -1,83 +1,58 @@
+        logger.info('Check whether it is time to withdraw my earnings')
+        const { fraction, sponsorshipAddresses } = await this.helper.getMyUnwithdrawnEarnings()
+        const safeUnwithdrawnEarningsFraction = this.penaltyLimitFraction! * this.withdrawLimitSafetyFraction / BigInt(ONE_ETHER)
+        logger.trace(` -> is ${fraction / BigInt(1e16)}% > ${safeUnwithdrawnEarningsFraction / BigInt(1e16)}% ?`)
```
Use `ONE_ETHER`

samigabor replied:
The 1e16 is 0.01 ether and that value is only used in those dev logs.

teogeb replied:
Easy to confuse to the full ether. So this is essentially just converting a value between 0-1 to a percentage (0-100):
```
logger.trace(` -> is ${fraction / ONE_ETHER * 100}% > ${safeUnwithdrawnEarningsFraction / ONE_ETHER * 100}% ?`)
```

samigabor replied:
Updated

teogeb commented on packages/broker/src/plugins/operator/OperatorPlugin.ts:
```diff
@@ -11,12 +11,16 @@ import { Signer } from '@ethersproject/abstract-signer'
 import { AnnounceNodeToContractHelper } from './AnnounceNodeToContractHelper'
 import { CONFIG_TEST } from 'streamr-client'
 
+export const DEFAULT_MAX_SPONSORSHIP_IN_WITHDRAW = 20 // max number of sponsorships to loop over before tx reverts
```
What `tx` this refers to? Could update the comment so that it it is understandable without knowning the implementation details?

samigabor replied:
Updated

teogeb commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,63 @@
+    private readonly abortController: AbortController
+    
+    // public access modifier for tests 
+    readonly helper: MaintainOperatorValueHelper
```
Is it possible to write the test wihtout exposing this? One possible solution would be to implement same pattern which we use `VoteOnSuspectNodeService.test.ts` which deploys an isolated environment?

samigabor replied:
It's not possible to deploy a new dev env for these tests (using streamrEnvDeployer) because we also need to query the existing subgraph. The alternative would be to build a new subgraph image on the fly for the newly generated addresses

teogeb commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,63 @@
+            return
+        }
+        logger.info('Check unwithdrawn earnings', { randomOperatorAddress })
+        return this.checkUnwithdrawnEarningsOf(randomOperatorAddress)
```
```suggestion
        await this.checkUnwithdrawnEarningsOf(randomOperatorAddress)
```

teogeb commented on packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    private async getPenaltyLimitFraction(): Promise<bigint> {
+        if (!this.penaltyLimitFractionCached) {
```
```suggestion
        if (!this.penaltyLimitFractionCached !== undefined) {
```

samigabor replied:
Updated

teogeb commented on packages/broker/test/integration/plugins/operator/MaintainOperatorValueHelper.test.ts:
```diff
@@ -0,0 +1,51 @@
+const chainConfig = config.dev2
+const theGraphUrl = `http://${STREAMR_DOCKER_DEV_HOST}:8800/subgraphs/name/streamr-dev/network-subgraphs`
+
+const STREAM_CREATION_KEY = '0xb1abdb742d3924a45b0a54f780f0f21b9d9283b231a0a0b35ce5e455fa5375e7'
```
Maybe better name if we don't create streams with this key?

samigabor replied:
Updated

teogeb commented on packages/broker/test/integration/plugins/operator/MaintainOperatorValueService.test.ts:
```diff
@@ -1,125 +1,78 @@
+        // 5 DATA = 50% of 10 DATA, 50% is the "safe threshold" (we don't want to wait all the way to the limit, lest we be late)
+        // 10 DATA = 10% of 100 DATA, 10% is the penalty limit (sum of unwithdrawn earnings may not exceed it)
+        // 100 DATA = pool value
+        await service.start()
```
Are these comments about line 70-71? Or how they relate to `service.start()`

samigabor replied:
The idea was to explain through numbers what will happen when the service starts (what's the penalty limit and the safe threshold for the operator)

teogeb replied:
I don't still understand the comment... Maybe clarify a bit?

samigabor replied:
I see your point. Maybe those comments aren't describing much related to start. Removed them.

teogeb commented on packages/broker/test/integration/plugins/operator/OperatorPlugin.test.ts:
```diff
@@ -23,7 +23,7 @@ describe('OperatorPlugin', () => {
             extraPlugins: {
                 operator: {
-                    operatorContractAddress: '0x4A5C0EC07F7ddBd4B6050181638e24b0153991b2'
+                    operatorContractAddress: '0x139dfa493a45364b598f2f98e504192819082c85'
```
Why this is changed?

samigabor replied:
It was an outdated address (not anymore part of the config) after the fast-chain started to be used. Not sure how ci skipped it and got into streamr-1.0 (no impl for the operator plugin service was added in this PR). The new address is the one that's now always deployed into the fast-chain container.

teogeb commented on packages/broker/test/integration/plugins/operator/OperatorValueBreachWatcher.test.ts:
```diff
@@ -0,0 +1,84 @@
+        const operatorValueBreachWatcher = new OperatorValueBreachWatcher(watcherConfig)
+
+        const poolValueBeforeWithdraw = await operatorContract.getApproximatePoolValue()
+        const allowedDifference = poolValueBeforeWithdraw.div('10').toBigInt()
```
Why `10`? Can we read/calculate that value from somewhere?

samigabor replied:
It's the 10% penalty limit. Removed the hard-coded value and retrieved from the config

teogeb commented on packages/broker/test/integration/plugins/operator/deploySponsorshipContract.ts:
```diff
@@ -1,39 +1,50 @@
+    const sponsorshipFactory = new Contract(contracts.SponsorshipFactory, sponsorshipFactoryABI, deployer) as unknown as SponsorshipFactory
+    const streamRegistry = new Contract(contracts.StreamRegistry, streamRegistryABI, deployer) as unknown as StreamRegistry
+
+    const streamExists = await streamRegistry.exists(streamId)
```
Why this check is added? The core functionality of this method is to deploy a sponsorship, and it is maybe ok if this test helper function relies that it is passed valid data. (As we don't need to provide extra dependencies just for checking the dependencies.)

samigabor replied:
Not needed as we can indeed assume the factory will correctly deploy the stream. Removed it.

samigabor committed:
refactor: add stricter type for the operators response from the subgraph query

samigabor committed:
docs: make comment more descriptive

samigabor committed:
refactor: add more restrictive check

samigabor committed:
refactor: Update var name

samigabor committed:
refactor: remove hard-coded penalty limit and get it from the config

samigabor committed:
refactor: remove stream existence check on the deploy sponsorship utility since it can be assumed stream registry will correctly create the stream

samigabor committed:
Update packages/broker/src/plugins/operator/OperatorValueBreachWatcher.ts

Co-authored-by: Teo Gebhard <teo@streamr.com>

teogeb commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,122 @@
+            }
+        }
+        const parseItems = (response: { operators: { id: string }[] }) => {
+            if (!response.operators) {
```
If no operators is found, we receive an empty array from TheGraph:
```
{
  "data": {
    "operators": []
  }
}
```

We can just return it, and no need to have this extra check?

samigabor replied:
Updated

samigabor committed:
docs: remove comments describing internal details from start

samigabor committed:
refactor: use one ether const for log message

samigabor committed:
refactor: remove unnecessary check for the returned operators array from the subgraph

samigabor committed:
Merge branch 'streamr-1.0' into NET-537_operator_value_breach_watcher

teogeb commented on packages/broker/test/integration/plugins/operator/deploySponsorshipContract.ts:
```diff
@@ -1,39 +1,45 @@
 export async function deploySponsorship(
     // eslint-disable-next-line max-len
-    chainConfig: { contracts: { SponsorshipFactory: string, SponsorshipStakeWeightedAllocationPolicy: string, SponsorshipDefaultLeavePolicy: string, SponsorshipVoteKickPolicy: string } },
+    chainConfig: { contracts: { SponsorshipFactory: string, SponsorshipStakeWeightedAllocationPolicy: string, SponsorshipDefaultLeavePolicy: string, SponsorshipVoteKickPolicy: string, StreamRegistry: string } },
```
Why this is added? It is not used

samigabor replied:
Was used previously and got left there by mistake. Removed it.

teogeb commented on packages/broker/test/integration/plugins/operator/deploySponsorshipContract.ts:
```diff
@@ -1,39 +1,45 @@
-    const sponsorshipFactory =
-        new Contract(chainConfig.contracts.SponsorshipFactory, sponsorshipFactoryABI, deployer) as unknown as SponsorshipFactory
-    // console.log("deployer balance", await deployer.getBalance())
+    logger.debug('Stream to be sponsored', { streamId })
```
```suggestion
```

teogeb commented on packages/broker/test/integration/plugins/operator/deploySponsorshipContract.ts:
```diff
@@ -1,39 +1,45 @@
 ): Promise<Sponsorship> {
+    const { contracts } = chainConfig
+    logger.debug('deploySponsorship', { contracts, deployerBalance: await deployer.getBalance() })
+
```
The deployer should always have enough balance in the test environment? (Or if not in some test, the test will fail)

```suggestion
```

samigabor replied:
Removed the log. The contracts const is needed further down

teogeb commented on packages/broker/test/integration/plugins/operator/operatorValueUtils.ts:
```diff
@@ -0,0 +1,19 @@
+    for (const e of earnings) {
+        unwithdrawnEarnings += e.toBigInt()
+    }
+    logger.debug(`Total unwithdrawn earnings: ${formatEther(unwithdrawnEarnings.toString())} (t = ${Date.now()})`)
```
Each log message will usually include a timestamp automatically

```suggestion
    logger.debug(`Total unwithdrawn earnings: ${formatEther(unwithdrawnEarnings.toString())}`)
```

teogeb commented on packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts:
```diff
@@ -1,33 +1,117 @@
+        }
+        const parseItems = (response: { operators: { id: string }[] }) => response.operators
+        this.theGraphClient.updateRequiredBlockNumber(requiredBlockNumber)
+        const queryResult = this.theGraphClient.queryEntities<{ id: string }>(createQuery, parseItems)
```
The default parseItems works ok as there is single root level item which is an array:
https://github.com/streamr-dev/network/blob/NET-537_operator_value_breach_watcher/packages/utils/src/TheGraphClient.ts#L61-L62

```suggestion
        const queryResult = this.theGraphClient.queryEntities<{ id: string }>(createQuery)
```

samigabor replied:
Cool, thanks for pointing that out

samigabor committed:
fix: remove unused param

samigabor committed:
fix: remove deployer balance log since it's not relevant

samigabor committed:
Update packages/broker/test/integration/plugins/operator/deploySponsorshipContract.ts

Co-authored-by: Teo Gebhard <teo@streamr.com>

samigabor committed:
Update packages/broker/test/integration/plugins/operator/operatorValueUtils.ts

Co-authored-by: Teo Gebhard <teo@streamr.com>

samigabor committed:
Update packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts

Co-authored-by: Teo Gebhard <teo@streamr.com>

samigabor committed:
fix: lint error for unused vars
